core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java

```diff
@@ -273,35 +227,11 @@ public void initialize(String name, Map<String, String> unresolved) {
       this.endpoints = ImmutableSet.copyOf(config.endpoints());
     }
 
-    this.sessions = newSessionCache(mergedProps);
-    this.tableSessions = newSessionCache(mergedProps);
-    this.keepTokenRefreshed =
-        PropertyUtil.propertyAsBoolean(
-            mergedProps,
-            OAuth2Properties.TOKEN_REFRESH_ENABLED,
-            OAuth2Properties.TOKEN_REFRESH_ENABLED_DEFAULT);
     this.client = clientBuilder.apply(mergedProps);
     this.paths = ResourcePaths.forCatalogProperties(mergedProps);
 
-    String token = mergedProps.get(OAuth2Properties.TOKEN);
-    this.catalogAuth =
-        new AuthSession(
-            baseHeaders,
-            AuthConfig.builder()
-                .credential(credential)
-                .scope(scope)
-                .oauth2ServerUri(oauth2ServerUri)
-                .optionalOAuthParams(optionalOAuthParams)
-                .build());
-    if (authResponse != null) {
-      this.catalogAuth =
-          AuthSession.fromTokenResponse(
-              client, tokenRefreshExecutor(name), authResponse, startTimeMillis, catalogAuth);
-    } else if (token != null) {
-      this.catalogAuth =
-          AuthSession.fromAccessToken(
-              client, tokenRefreshExecutor(name), token, expiresAtMillis(mergedProps), catalogAuth);
-    }
+    authManager.initialize(name, client, mergedProps);
```

Comment:
I feel like we should have separate AuthManager instances, not double-initialize.

Reply:
I agree that, on the paper, that would look nicer, but the devil is in the details: consider for instance the case where `credential` is provided. The auth manager would have to exchange the provided credentials for an access token using the `client_credentials` grant. Now, if we have two distinct auth managers, _both_ would have to perform this roundtrip since they would not be able to share any state; this would result in two tokens being requested in a row.

Do you think that this extra network roundtrip would be acceptable? If so, then yes, we can split in two instances.

Reply:
Conceptually, the client credential flow is the riskiest part because it puts long-term credentials on the wire. Ideally, I think the user should be able to control where and when that happens.

For example, it may be desirable for the Spark driver to execute the client credentials flow (once) and only send the resulting access and refresh tokens to worker nodes.

I'm not sure how feasible that is in the current state of the code, but I'm bringing this up for consideration as I think it is an important feature.

Reply:
OK, I was able to improve things a bit. 

I'm not using two instances because I think it's important to reuse the authentication data obtained from the interaction with the config endpoint; but, I removed the double initialization, that indeed felt a bit awkward. 

I also removed the double call to the same `catalogSession` method; there is now a separate `preConfigSession` method that is meant to return a temporary, short-lived session for contacting the config endpoint only. 

This is in commit 4abc8dc. Let me know what you think!

Reply:
I guess I was not thinking clearly when I made my comment above. I think in general it would be good to perhaps avoid keeping the client secret in memory after the initial grant and then use the refresh token only... but I think this PR moves in the right direction :)